sw/device/lib/drivers/gpio/gpio.c

```diff
@@ -1,213 +1,624 @@
+ * @brief get some bitfield of a 32b reg.
+ * @param reg the register that bitwise field selects from.
+ * @param mask the mask w/o offset.
+ * @param sel this the offset used for mask and value.
+ */
+static inline uint32_t getBitfield( uint32_t reg, uint32_t  mask, uint8_t sel )
+{
+    uint32_t ret_val = 0;
+    ret_val = reg & ( mask << sel );
+    ret_val = ret_val >> sel;
+    return ret_val;
+}
+
+/****************************************************************************/
+/**                                                                        **/
+/*                           EXPORTED VARIABLES                             */
+/**                                                                        **/
+/****************************************************************************/
+
+/****************************************************************************/
+/**                                                                        **/
+/*                            GLOBAL VARIABLES                              */
+/**                                                                        **/
+/****************************************************************************/
+
+/****************************************************************************/
+/**                                                                        **/
+/*                           EXPORTED FUNCTIONS                             */
+/**                                                                        **/
+/****************************************************************************/
+
+gpio_result_t gpio_config (gpio_cfg_t cfg){
+    // /* check that passed arg is not empty*/
+    // if (cfg == NULL)
+    //     return GpioError;
+    /* check that pin is in acceptable range. */
+    if (cfg.pin > (MAX_PIN-1) || cfg.pin < 0)
+        return GpioError;
+    /* reset pin coniguration first.*/
+    gpio_reset (cfg.pin);
+    /* check mode. */
+    if ((cfg.mode < GpioModeIn) || (cfg.mode > GpioModeoutOpenDrain1))
+        return GpioError;
+    /* set mode. */
+    gpio_set_mode (cfg.pin, cfg.mode);
+    /* if input sampling is enabled */
+    if (cfg.en_input_sampling == true)
+        gpio_en_input_sampling (cfg.pin);
+    /* if interrupt is enabled. Also after enabling check for error */
+    if (cfg.en_intr == true)
+        if (gpio_intr_en (cfg.pin, cfg.intr_type) == GpioError)
+            return GpioError;
+    return GpioOk;
+}
+
+gpio_result_t gpio_set_mode (gpio_pin_number_t pin, gpio_mode_t mode)
+{
+    if (pin >= 0 && pin < 16)
+    {
+        setBitfield(&(gpio_peri->GPIO_MODE0), mode, 0b11, 2*pin);
+        return GpioOk;
+    }
+    else if (pin >= 16 && pin <MAX_PIN)
+    {
+        setBitfield(&(gpio_peri->GPIO_MODE1), mode, 0b11, 2*(pin-16));
+        return GpioOk;
+    }
+    else
+    {
+        return GpioError;
+    }
+}
+
+gpio_result_t gpio_en_input_sampling (gpio_pin_number_t pin)
+{
+    if (pin >= 0 && pin < MAX_PIN)
```

Comment:
can we set this also for output pins?? if not, we should add some sanity check?

Reply:
you are right, we can do the sanity check to not enable it by mistake for output pins. When it calls through gpio_config sanity check is done inside the function, however, when it is called separately it may happen, so yes good to add!